File: python/백준/구현/21608 상어 초등학교.py
import sys

input = sys.stdin.readline
 
N = int(input())

like_dict = {}
order = []

for _ in range(N * N):
    data = list(map(int, input().split()))
    student = data[0]
    likes = data[1:]
    like_dict[student] = likes
    order.append(student)

seat = [[0] * N for _ in range(N)]

directions = [(0,1), (1,0), (-1,0), (0,-1)]  # 우, 하, 상, 좌 (순서는 상관 X)

def find_best_seat(student):
    likes = like_dict[student]
    best = None  # (like_cnt, empty_cnt, r, c)
    
    for r in range(N):
        for c in range(N):
            if seat[r][c] != 0:
                continue

            like_count = 0
            empty_count = 0

            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if 0 <= nr < N and 0 <= nc < N:
                    if seat[nr][nc] in likes:
                        like_count += 1
                    elif seat[nr][nc] == 0:
                        empty_count += 1
            
            if best is None:
                best = (like_count, empty_count, r, c)
            else:
                if like_count > best[0]:
                    best = (like_count, empty_count, r, c)
                elif like_count == best[0]:
                    if empty_count > best[1]:
                        best = (like_count, empty_count, r, c)
                    elif empty_count == best[1]:
                        if r < best[2] or (r == best[2] and c < best[3]):
                            best = (like_count, empty_count, r, c)

    return best[2], best[3]


for student in order:
    r, c = find_best_seat(student)
    seat[r][c] = student

score_table = [0, 1, 10, 100, 1000]
answer = 0

for r in range(N):
    for c in range(N):
        student = seat[r][c]
        likes = like_dict[student]
        like_cnt = 0
        
        for dr, dc in directions:
            nr, nc = r + dr, c + dc
            if 0 <= nr < N and 0 <= nc < N:
                if seat[nr][nc] in likes:
                    like_cnt += 1

        answer += score_table[like_cnt]

print(answer)

File: cpp/백준/구현/21608 상어 초등학교.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <tuple>
#include <algorithm>
using namespace std;

int N;
unordered_map<int, vector<int>> like_dict;
vector<vector<int>> seat;

int directions[4][2] = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};

pair<int, int> find_best_seat(int student) {
    const vector<int>& likes = like_dict[student];
    bool has_best = false;
    tuple<int, int, int, int> best;

    for (int r = 0; r < N; ++r) {
        for (int c = 0; c < N; ++c) {
            if (seat[r][c] != 0) continue;

            int like_count = 0;
            int empty_count = 0;

            for (int d = 0; d < 4; ++d) {
                int nr = r + directions[d][0];
                int nc = c + directions[d][1];
                if (nr >= 0 && nr < N && nc >= 0 && nc < N) {
                    int val = seat[nr][nc];
                    if (find(likes.begin(), likes.end(), val) != likes.end()) {
                        like_count++;
                    } else if (val == 0) {
                        empty_count++;
                    }
                }
            }

            tuple<int, int, int, int> cur = {like_count, empty_count, r, c};
            if (!has_best) {
                best = cur;
                has_best = true;
            } else {
                auto [lc, ec, br, bc] = best;
                if (like_count > lc) {
                    best = cur;
                } else if (like_count == lc) {
                    if (empty_count > ec) {
                        best = cur;
                    } else if (empty_count == ec) {
                        if (r < br || (r == br && c < bc)) {
                            best = cur;
                        }
                    }
                }
            }
        }
    }

    return {get<2>(best), get<3>(best)};
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N;

    vector<int> order;
    for (int i = 0; i < N * N; ++i) {
        int student;
        cin >> student;
        vector<int> likes(4);
        for (int j = 0; j < 4; ++j) cin >> likes[j];
        like_dict[student] = likes;
        order.push_back(student);
    }

    seat.assign(N, vector<int>(N, 0));

    for (int student : order) {
        auto [r, c] = find_best_seat(student);
        seat[r][c] = student;
    }

    int score_table[] = {0, 1, 10, 100, 1000};
    int answer = 0;

    for (int r = 0; r < N; ++r) {
        for (int c = 0; c < N; ++c) {
            int student = seat[r][c];
            const vector<int>& likes = like_dict[student];
            int like_cnt = 0;

            for (int d = 0; d < 4; ++d) {
                int nr = r + directions[d][0];
                int nc = c + directions[d][1];
                if (nr >= 0 && nr < N && nc >= 0 && nc < N) {
                    if (find(likes.begin(), likes.end(), seat[nr][nc]) != likes.end()) {
                        like_cnt++;
                    }
                }
            }

            answer += score_table[like_cnt];
        }
    }

    cout << answer << "\n";
    return 0;
}
